torch/csrc/jit/passes/alias_analysis.cpp

```diff
@@ -927,20 +931,50 @@ void AliasDb::move(Node* toMove, Node* movePoint, MoveSide moveSide) {
   }
 }
 
+c10::optional<const Node*> AliasDb::getLastWildcard() const {
+  auto it = std::max_element(
+      wildcardNodes_.cbegin(),
+      wildcardNodes_.cend(),
+      [this](const Node* a, const Node* b) { return isBeforeSameGraph(a, b); });
+  if (it != wildcardNodes_.end()) {
+    return *it;
+  } else {
+    return c10::nullopt;
+  }
+}
+
 bool AliasDb::hasUntrackedEffects(Node* node) const {
   bool touchesWildcard = false;
-  if (!wildcardNodes_.empty()) {
-    auto lastWildcard = *wildcardNodes_.begin();
-    for (const auto wildcard : wildcardNodes_) {
-      if (wildcard->isAfter(lastWildcard)) {
-        lastWildcard = wildcard;
-      }
-    }
+  if (const auto lastWildcard = getLastWildcard()) {
     touchesWildcard = hasWrites(node) &&
-        (node->isBefore(lastWildcard) || node == lastWildcard);
+        (isBeforeSameGraph(node, *lastWildcard) || node == *lastWildcard);
   }
 
   return writesToInputAlias(node) || touchesWildcard;
 }
+
+// Nodes must be in the same graph in order to do `isBefore` or `isAfter`. This
+// traverses the subgraph "chain" upward until we find two nodes that share an
+// owning graph.
+bool AliasDb::isBeforeSameGraph(const Node* a, const Node* b) const {
+  auto lhs = a;
```

Comment:
So this is iteratively moving b all the way up graph and checking if it is in the same chain as a. Then it moves a up one level, and starts b at its original position. This is O(n^2), which doesn't super matter in our current state since we don't tend to have many nested graphs and n is small.

Alternative: check A and B's depth from the root, move up either node until they are at the same depth. 
Iteratively test if they are in the same graph, and if not move them up. This would be O(n) & IMO be a little easier to follow.

Reply:
Yeah it's n^2 in the depth of subgraph. Since the depth right now is never more than like 2, I think it's okay. 

If we start breaking out function calls using a resolution table instead of inlining them, we will have some sort of "aliasing summary" for a graph anyway, so we will not have to search subgraphs at all.

Reply:
it's more that I think the other algorithm is more readable since it doesn't have a double nesting of loops and reads from top to bottom. 

Could you at add a comment about why it's okay we're doing the O(n^2) algorithm instead of the O(n) ?